server/users/user-settings.model.ts

```diff
@@ -0,0 +1,33 @@
+import { DomainModel } from "../core/domain.model"
+import { UserSettingModel } from "./user-setting-model"
+export class UserSettingsModel implements DomainModel {
+
+	public readonly id: number | undefined
+	public readonly userId: number | undefined
+	public readonly receiveTopRated: boolean
+	public readonly receiveTopRatedSetting: UserSettingModel
+
+	constructor(
+		id: number | undefined,
+		userId: number | undefined,
+        receiveTopRated: boolean = false
+	) {
+		this.id = id
+		this.userId = userId
+
+		this.receiveTopRated = receiveTopRated
+		this.receiveTopRatedSetting = new UserSettingModel("Receive notifications about top-rated sport events", receiveTopRated, "receive_top_rated")
```

Comment:
I don't like the fact that we're constructing the `UserSettingModel` here. I think factory should do it and pass it down to this class. We can use `UserSettingsFactory` or even `UserFactory`, we probably don't need specific `UserSettingsModelFactory` for a class as simple as this.

Biggest problem with this is that we're hardcoding database column name here, and domain models shouldn't be aware of how they're read from or saved to the database. We can introduce an enum for all possible settings, and then factories and services can use it map from enum value to database column.

Also when you do it, try to reuse `UserSettingsAttributes` interface generated from database to make sure there's some type safety when setting database column name.

Reply:
You can use `Pick` and `keyof` to achieve it:

![image](https://user-images.githubusercontent.com/3044893/139761275-afd14ed7-c645-4213-8040-cd5c0b0c5096.png)

Reply:
Not sure I understand this completely. Say I now create the object 
```
new UserSettingModel("Receive notifications about (...)
```
in `UserFactory`. Where should I perform these checks, using `Pick` and `keyof`, since I presume I shouldn't be modifying the generated database/* models?

Reply:
Refactored in: https://github.com/mbrizic/ratethegame/pull/28/commits/0713f38e2625f1f98f02c12982914cdc2f903b6e
without any validation, now hardcoded in `UserFactory`

Reply:
Validation added in: https://github.com/mbrizic/ratethegame/pull/28/commits/3ba2dd83259fe0abc82ee85f05690d758c0d3347